pkg/providers/instance/vminstance.go

```diff
@@ -66,6 +66,19 @@ const (
 	cseNameLinux   = "cse-agent-karpenter"
 )
 
+// errorCodeForMetrics extracts a stable Azure error code for metric labeling when possible, preventing overcardinality in metrics with failures.
+func errorCodeForMetrics(err error) string {
```

Comment:
minor: Future item but I think we could upstream this to azure sdk for go extensions as:
`ErrorCode(err error) string` and just document that it returns a static string `Unknown` if the error is not of type ResponseError.